File: docs/JAMIA/comparisons/RPython/TM/21011777_isnan_on_mixed_list.py
# ID: SO#21011777
# Language: Python
# Bug class: Value type
# Runtime consequence: Crash
# In study: Yes
# Inclusion rationale: NaN mixed into list prevents clean removal — math.isnan fails on non-float elements. Mixed types.
# Reproduction status: Live
# Type system catch: `number | null` typed column; `isNaN` rejects non-number elements

import math
import numpy as np

countries = [float('nan'), 'USA', 'UK', 'France']

# BUG: math.isnan fails on string elements
try:
    cleaned = [x for x in countries if not math.isnan(x)]
    print(cleaned)
except TypeError as e:
    print(f"Crash (math.isnan): {e}")

# BUG: np.isnan also fails on mixed types
try:
    mask = np.isnan(countries)
    print(mask)
except TypeError as e:
    print(f"Crash (np.isnan): {e}")

File: docs/JAMIA/comparisons/RPython/TM/21011777_isnan_on_mixed_list.ts
/**
 * ID: SO#21011777
 * Language: Python
 * Bug class: Value type
 * Runtime consequence: Crash
 * In study: Yes
 * Inclusion rationale: NaN mixed into list prevents clean removal — math.isnan fails on non-float elements. Mixed types.
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";
import { printForeignResult, runForeign } from "../run-foreign.ts";

// Foreign reproduction (pandas) ──────────────────────────────────────────────

const foreignScript = `
import math

countries = [float('nan'), 'USA', 'UK', 'France']

cleaned = [x for x in countries if not math.isnan(x)]
`;

printForeignResult("python", runForeign("python", foreignScript));

// Tidy-TS equivalent ─────────────────────────────────────────────────────────

const df = createDataFrame([
  { country: null as string | null },
  { country: "USA" },
  { country: "UK" },
  { country: "France" },
]);

// @ts-expect-error — Type '(string | null)[]' is not assignable to type 'number[]'
s.mean(df.extract("country"));
